python-package/lightgbm/basic.py

```diff
@@ -296,7 +296,9 @@ def _data_from_pandas(data, feature_name, categorical_feature, pandas_categorica
             raise ValueError("DataFrame.dtypes for data must be int, float or bool.\n"
                              "Did not expect the data types in the following fields: "
                              + ', '.join(data.columns[bad_indices]))
-        data = data.values.astype('float')
+        data = data.values
+        if data.dtype != 'float':
```

Comment:
Thanks for this! I apologize in advance if you got to this PR because you experienced an out-of-memory error.

I think it's a good change, but I don't think this syntax is quite right. `pandas` DataFrames don't have a `.dtype`, they have an array of dtypes (one per column)

```python
import pandas as pd
df = pd.DataFrame({'a': ['1', '2', '3'], 'b': [4, 5, 6]})
df.dtype
```
will return

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jlamb/miniconda3/lib/python3.6/site-packages/pandas/core/generic.py", line 4376, in __getattr__
    return object.__getattribute__(self, name)
AttributeError: 'DataFrame' object has no attribute 'dtype'
```

where `.dtypes` will give you

```
df.dtypes
a    object
b     int64
dtype: object
```

I think you want something like this

```python
import numpy as np
all_float = all([col_type == np.dtype('float') for col_type in data.dtypes])
if not all_float:
    data = data.astype('float')
```

@StrikerRUS I'm surprised that this syntax error (`data.dtype`) wasn't caught by our CI. Have I made a mistake in my understanding of what `data` is?

Reply:
@jameslamb [`data.values` returns a numpy array](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.values.html), thus `data.values.dtype` is fine.

Reply:
😆 I read straight past the `.values`, haha. Sorry! Ok then I am good with this change! Will leave it for @StrikerRUS to review just in case, since I am not the main owner of the Python side for this project, but this seems good to me.

Also TIL that `'float' == dtype('float')` works! That's interesting.

```
import pandas as pd
df = pd.DataFrame({'a': [1.0, 2.0, 3.0], 'b': [4.0, 5.0, 6.0]})
print(df.values.dtype)
print(df.values.dtype == 'float')
```